src/main/java/com/google/api/generator/gapic/composer/grpc/GrpcServiceStubClassComposer.java

```diff
@@ -245,10 +255,171 @@ protected String getProtoRpcFullMethodName(Service protoService, Method protoMet
+      String pathTemplateName =
+          String.format("PATH_TEMPLATE_%s_%s", method.name().toUpperCase(), i);
+      TypeNode pathTemplateType =
+          TypeNode.withReference(ConcreteReference.withClazz(PathTemplate.class));
+      Variable pathTemplateVar =
+          Variable.builder().setType(pathTemplateType).setName(pathTemplateName).build();
+      Expr routingHeaderPatternExpr = VariableExpr.withVariable(pathTemplateVar);
+      VariableExpr pathTemplateVarExpr =
+          VariableExpr.builder()
+              .setVariable(pathTemplateVar)
+              .setIsDecl(true)
+              .setIsStatic(true)
+              .setIsFinal(true)
+              .setScope(ScopeNode.PRIVATE)
+              .build();
+      ValueExpr valueExpr =
+          ValueExpr.withValue(StringObjectValue.withValue(routingHeader.pattern()));
+      Expr pathTemplateExpr =
+          AssignmentExpr.builder()
+              .setVariableExpr(pathTemplateVarExpr)
+              .setValueExpr(
+                  MethodInvocationExpr.builder()
+                      .setStaticReferenceType(pathTemplateType)
+                      .setMethodName("create")
+                      .setArguments(valueExpr)
+                      .setReturnType(pathTemplateType)
+                      .build())
+              .build();
+      Statement pathTemplateClassVar = ExprStatement.withExpr(pathTemplateExpr);
+      classStatements.add(pathTemplateClassVar);
+      Expr valueOfExpr =
+          MethodInvocationExpr.builder()
+              .setStaticReferenceType(TypeNode.STRING)
+              .setMethodName("valueOf")
+              .setArguments(requestFieldGetterExpr)
+              .build();
+      MethodInvocationExpr addParamsMethodExpr =
+          MethodInvocationExpr.builder()
+              .setExprReferenceExpr(routingHeadersBuilderVarNonDeclExpr)
+              .setMethodName("add")
+              .setArguments(valueOfExpr, routingHeaderKeyExpr, routingHeaderPatternExpr)
+              .build();
+
+      IfStatement ifStatement =
+          IfStatement.builder()
+              .setConditionExpr(
+                  fieldValuesNotNullConditionExpr(
+                      requestVarExpr, routingHeader.getDescendantFieldNames()))
+              .setBody(ImmutableList.of(ExprStatement.withExpr(addParamsMethodExpr)))
+              .build();
+      bodyStatements.add(ifStatement);
+    }
+
+    return MethodInvocationExpr.builder()
+        .setExprReferenceExpr(routingHeadersBuilderVarNonDeclExpr)
+        .setMethodName("build")
+        .setReturnType(returnType)
+        .build();
+  }
+
+  private Expr fieldValuesNotNullConditionExpr(
+      VariableExpr requestVarExpr, List<String> fieldNames) {
+    MethodInvocationExpr.Builder requestFieldGetterExprBuilder =
+        MethodInvocationExpr.builder().setExprReferenceExpr(requestVarExpr);
+    Expr fieldValuesNotNullExpr = null;
+    for (int i = 0; i < fieldNames.size(); i++) {
```

Comment:
it does not seem like there are any other usages for `i` except `get(i)`, can we just switch to foreach loop then?

Reply:
Updated the logic to not adding null check for the last field, so we need the index now to skip the last one.